README.md
dmtools <img src='man/figures/logo.png' align="right" height="170" />
=====================================================================

[![CRAN
status](https://www.r-pkg.org/badges/version/dmtools)](https://CRAN.R-project.org/package=dmtools)
[![Build
Status](https://travis-ci.com/chachabooms/dmtools.svg?token=pmH5ZxVz4xaZTjx5TDKs&branch=master)](https://travis-ci.com/chachabooms/dmtools)
[![codecov](https://codecov.io/gh/KonstantinRyabov/dmtools/branch/master/graph/badge.svg?token=AEKUFWUUXZ)](https://codecov.io/gh/KonstantinRyabov/dmtools)

Installation
------------

    install.packages("dmtools")

    # dev-version
    devtools::install_github("KonstantinRyabov/dmtools")

    library(dmtools)

Overview
--------

For checking the dataset from EDC in clinical trials. Notice, your
dataset should have a postfix( \_V1 ) or a prefix( V1\_ ) in the names
of variables. Column names should be unique.

-   `date()` - create object date to check dates in the dataset
-   `lab()` - create object lab to check lab reference range
-   `short()` - create object short to reshape the dataset in a tidy
    view.
-   `check()` - check objects
-   `get_result()` - get the final result of object
-   `choose_test()` - filter the final result of `check()`
-   `rename_dataset()` - rename the dataset

Usage
-----

For example, you want to check laboratory values, you need to create the
excel table like in the example.

-   AGELOW - number, &gt;= number
-   AGEHIGH - if none, type Inf, &lt;= number  
-   SEX - for both sex, use `|`
-   LBTEST - What was the lab test name? (can be any convenient name for
    you)
-   LBORRES\* - What was the result of the lab test?
-   LBNRIND\* - How \[did/do\] the reported values compare within the
    \[reference/normal/expected\] range?
-   LBORNRLO - What was the lower limit of the reference range for this
    lab test, &gt;=
-   LBORNRHI - What was the high limit of the reference range for this
    lab test, &lt;=

\*column names without prefix or postfix

<table>
<caption>lab reference ranges</caption>
<thead>
<tr class="header">
<th style="text-align: right;">AGELOW</th>
<th style="text-align: right;">AGEHIGH</th>
<th style="text-align: left;">SEX</th>
<th style="text-align: left;">LBTEST</th>
<th style="text-align: left;">LBORRES</th>
<th style="text-align: left;">LBNRIND</th>
<th style="text-align: left;">LBORNRLO</th>
<th style="text-align: left;">LBORNRHI</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: right;">18</td>
<td style="text-align: right;">45</td>
<td style="text-align: left;">f|m</td>
<td style="text-align: left;">Glucose</td>
<td style="text-align: left;">GLUC</td>
<td style="text-align: left;">GLUC_IND</td>
<td style="text-align: left;">3.9</td>
<td style="text-align: left;">5.9</td>
</tr>
<tr class="even">
<td style="text-align: right;">18</td>
<td style="text-align: right;">45</td>
<td style="text-align: left;">m</td>
<td style="text-align: left;">Aspartate transaminase</td>
<td style="text-align: left;">AST</td>
<td style="text-align: left;">AST_IND</td>
<td style="text-align: left;">0</td>
<td style="text-align: left;">42</td>
</tr>
<tr class="odd">
<td style="text-align: right;">18</td>
<td style="text-align: right;">45</td>
<td style="text-align: left;">f</td>
<td style="text-align: left;">Aspartate transaminase</td>
<td style="text-align: left;">AST</td>
<td style="text-align: left;">AST_IND</td>
<td style="text-align: left;">0</td>
<td style="text-align: left;">39</td>
</tr>
</tbody>
</table>

<table>
<caption>dataset</caption>
<thead>
<tr class="header">
<th style="text-align: left;">ID</th>
<th style="text-align: left;">AGE</th>
<th style="text-align: left;">SEX</th>
<th style="text-align: left;">GLUC_V1</th>
<th style="text-align: left;">GLUC_IND_V1</th>
<th style="text-align: left;">AST_V2</th>
<th style="text-align: left;">AST_IND_V2</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">01</td>
<td style="text-align: left;">19</td>
<td style="text-align: left;">f</td>
<td style="text-align: left;">5.5</td>
<td style="text-align: left;">norm</td>
<td style="text-align: left;">30</td>
<td style="text-align: left;">norm</td>
</tr>
<tr class="even">
<td style="text-align: left;">02</td>
<td style="text-align: left;">20</td>
<td style="text-align: left;">m</td>
<td style="text-align: left;">4.1</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">48</td>
<td style="text-align: left;">norm</td>
</tr>
<tr class="odd">
<td style="text-align: left;">03</td>
<td style="text-align: left;">22</td>
<td style="text-align: left;">m</td>
<td style="text-align: left;">9.7</td>
<td style="text-align: left;">norm</td>
<td style="text-align: left;">31</td>
<td style="text-align: left;">norm</td>
</tr>
</tbody>
</table>

    # "norm" and "no" it is an example, necessary variable for the estimate, get from the dataset
    refs <- system.file("labs_refer.xlsx", package = "dmtools")
    obj_lab <- lab(refs, ID, AGE, SEX, "norm", "no")
    obj_lab <- obj_lab %>% check(df)

    # ok - analysis, which has a correct estimate of the result
    obj_lab %>% choose_test("ok")
    #>   ID AGE SEX                 LBTEST LBTESTCD VISIT LBORNRLO LBORNRHI LBORRES
    #> 1 01  19   f                Glucose     GLUC   _V1      3.9      5.9     5.5
    #> 2 01  19   f Aspartate transaminase      AST   _V2      0.0     39.0      30
    #> 3 03  22   m Aspartate transaminase      AST   _V2      0.0     42.0      31
    #>   LBNRIND RES_TYPE_NUM IND_EXPECTED
    #> 1    norm          5.5         norm
    #> 2    norm         30.0         norm
    #> 3    norm         31.0         norm

    # mis - analysis, which has an incorrect estimate of the result
    obj_lab %>% choose_test("mis")
    #>   ID AGE SEX                 LBTEST LBTESTCD VISIT LBORNRLO LBORNRHI LBORRES
    #> 1 02  20   m Aspartate transaminase      AST   _V2      0.0     42.0      48
    #> 2 03  22   m                Glucose     GLUC   _V1      3.9      5.9     9.7
    #>   LBNRIND RES_TYPE_NUM IND_EXPECTED
    #> 1    norm         48.0           no
    #> 2    norm          9.7           no

    # skip - analysis, which has an empty value of the estimate
    obj_lab %>% choose_test("skip")
    #>   ID AGE SEX  LBTEST LBTESTCD VISIT LBORNRLO LBORNRHI LBORRES LBNRIND
    #> 1 02  20   m Glucose     GLUC   _V1      3.9      5.9     4.1    <NA>
    #>   RES_TYPE_NUM IND_EXPECTED
    #> 1          4.1         norm

<table>
<caption>strange_dataset</caption>
<thead>
<tr class="header">
<th style="text-align: left;">ID</th>
<th style="text-align: left;">AGE</th>
<th style="text-align: left;">SEX</th>
<th style="text-align: left;">V1_GLUC</th>
<th style="text-align: left;">V1_GLUC_IND</th>
<th style="text-align: left;">V2_AST</th>
<th style="text-align: left;">V2_AST_IND</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">01</td>
<td style="text-align: left;">19</td>
<td style="text-align: left;">f</td>
<td style="text-align: left;">5,5</td>
<td style="text-align: left;">norm</td>
<td style="text-align: left;">&lt; 5</td>
<td style="text-align: left;">norm</td>
</tr>
<tr class="even">
<td style="text-align: left;">02</td>
<td style="text-align: left;">20</td>
<td style="text-align: left;">m</td>
<td style="text-align: left;">4,1</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">48</td>
<td style="text-align: left;">norm</td>
</tr>
<tr class="odd">
<td style="text-align: left;">03</td>
<td style="text-align: left;">22</td>
<td style="text-align: left;">m</td>
<td style="text-align: left;">9,7</td>
<td style="text-align: left;">norm</td>
<td style="text-align: left;">31</td>
<td style="text-align: left;">norm</td>
</tr>
</tbody>
</table>

    # dmtools can work with the dataset as strange_df
    # parameter is_post has value FALSE because a dataset has a prefix( V1_ ) in the names of variables
    obj_lab <- lab(refs, ID, AGE, SEX, "norm", "no", is_post = F)
    obj_lab <- obj_lab %>% check(strange_df)

    # dmtools can understand the value with a comma like 6,6 
    obj_lab %>% choose_test("ok")
    #>   ID AGE SEX                 LBTEST LBTESTCD VISIT LBORNRLO LBORNRHI LBORRES
    #> 1 01  19   f                Glucose     GLUC   V1_      3.9      5.9     5,5
    #> 2 03  22   m Aspartate transaminase      AST   V2_      0.0     42.0      31
    #>   LBNRIND RES_TYPE_NUM IND_EXPECTED
    #> 1    norm          5.5         norm
    #> 2    norm         31.0         norm

    # Notice, if dmtools can't understand the value of lab_vals e.g. < 5, it puts Inf in the RES_TYPE_NUM
    obj_lab %>% choose_test("mis")
    #>   ID AGE SEX                 LBTEST LBTESTCD VISIT LBORNRLO LBORNRHI LBORRES
    #> 1 01  19   f Aspartate transaminase      AST   V2_      0.0     39.0     < 5
    #> 2 02  20   m Aspartate transaminase      AST   V2_      0.0     42.0      48
    #> 3 03  22   m                Glucose     GLUC   V1_      3.9      5.9     9,7
    #>   LBNRIND RES_TYPE_NUM IND_EXPECTED
    #> 1    norm          Inf           no
    #> 2    norm         48.0           no
    #> 3    norm          9.7           no

    obj_lab %>% choose_test("skip")
    #>   ID AGE SEX  LBTEST LBTESTCD VISIT LBORNRLO LBORNRHI LBORRES LBNRIND
    #> 1 02  20   m Glucose     GLUC   V1_      3.9      5.9     4,1    <NA>
    #>   RES_TYPE_NUM IND_EXPECTED
    #> 1          4.1         norm


NEWS.md
# dmtools 0.2.6
* fixed LBNDIND -> LBNRIND
* check if age or sex is null in `lab()`
* delete `check_sites()` and `test_sites()`
* delete parameter `clsig` in `lab()`
* add progress bar

# dmtools 0.2.5

* add functions for MedDRA API
* `short()` don't stop if an error happens
* CDASH
* delete `wbc()`

# dmtools 0.2.4

* add `short()`
* add `get_result()`
* add parameter `name_to_find` for `lab()`, `wbc()`
* delete parameter `test` and add paramater `func` for `test_sites()`
* add example for `rename_dataset()`

# dmtools 0.2.3

* add `test_sites` for choose test in the different sites
* add `get_date` for the date object

## Bug fixes

* wbc checks non-numeric arguments

# dmtools 0.2.2

* add opportunity for check sites
* delete depended and filtered columns

## Bug fixes

* parameter for clinical significant estimate

# dmtools 0.2.1

* add <- -> for dates, which is out
* add function for calculation WBCs count
* OOP structure

# dmtools 0.1.4

* throw error if the final result result is empty
* throw warning if don`t find item in file

## Bug fixes

* parameter for sex: substring in regex e.g. "male" in "female"

# dmtools 0.1.3

* add a parameter for the clinical significant estimate

# dmtools 0.1.2

* add a function for dates validation

# dmtools 0.1.1

* first public release of the package


R/lab.R
#' Create object lab
#'
#' @param file A character scalar. Path to the laboratory's reference in the excel table.
#' @param is_post A logical scalar, default is TRUE. True is postfix, otherwise, prefix.
#' @param id A column name of the subject id in the dataset, without quotes.
#' @param age A column name of the subject age in the dataset, without quotes.
#' @param sex A column name of the subject sex in the dataset, without quotes.
#' @param normal A normal estimate, for example, "NORMAL".
#' @param abnormal An abnormal estimate, for example, "ABNORMAL".
#' @param name_to_find A character scalar. For search prefixes or postfixes, default is "LBNRIND".
#'
#' @return The object lab.
#' @export
#'
#' @examples
#' obj_lab <- lab("lab_refer.xlsx", ID, AGE, SEX, 1, 2)
#' obj_lab <- lab("lab_refer.xlsx", ID, AGE, SEX, "NORMAL", "ABNORMAL")
#' obj_lab <- lab("lab_refer.xlsx", ID, AGE, SEX, "norm", "no", FALSE)
lab <- function(file,
                id,
                age,
                sex,
                normal,
                abnormal,
                is_post = T,
                name_to_find = "LBNRIND") {
  obj <- list(
    file = file,
    id = dplyr::enquo(id),
    age = dplyr::enquo(age),
    sex = dplyr::enquo(sex),
    normal = normal,
    abnormal = abnormal,
    is_post = is_post,
    name_to_find = name_to_find,
    bond = "_"
  )

  class(obj) <- "lab"
  obj
}

#' Filter the final result of the object lab
#'
#' @param obj An object. Class lab.
#' @param group_id A logical scalar, default is TRUE.True is grouped by id, otherwise, it isn't grouped.
#' @param test A character scalar. Parameters, which use to filter the final dataset, default: "mis":
#'                                      "ok" - analysis, which has a correct estimate of the result,
#'                                      "mis" - analysis, which has an incorrect estimate of the result,
#'                                      "skip" - analysis, which has an empty value of the estimate,
#'                                      "null" - analysis, which has an empty result and value of the estimate.
#'
#'
#' @return The filtered dataset by a value of \code{test}.
#' @export
#'
#' @examples
#' ID <- c("01", "02", "03")
#' SITE <- c("site 01", "site 02", "site 03")
#' AGE <- c("19", "20", "22")
#' SEX <- c("f", "m", "f")
#' GLUC_V1 <- c(5.5, 4.1, 9.7)
#' GLUC_IND_V1 <- c("norm", "no", "cl")
#' AST_V2 <- c("30", "48", "31")
#' AST_IND_V2 <- c(NA, "norm", "norm")
#'
#' df <- data.frame(
#'   ID, SITE, AGE, SEX,
#'   GLUC_V1, GLUC_IND_V1,
#'   AST_V2, AST_IND_V2,
#'   stringsAsFactors = FALSE
#' )
#'
#' refs <- system.file("labs_refer.xlsx", package = "dmtools")
#' obj_lab <- lab(refs, ID, AGE, SEX, "norm", "no")
#'
#' obj_lab <- check(obj_lab, df)
#' choose_test(obj_lab, "mis")
choose_test.lab <- function(obj, test = "mis", group_id = T) {
  result <- obj %>% get_result(group_id)

  # filter final dataset
  if (test == "mis") {
    result <-
      result %>%
      dplyr::filter(.data$IS_RIGHT == F) %>%
      dplyr::select(-.data$IS_RIGHT)
  } else if (test == "ok") {
    result <-
      result %>%
      dplyr::filter(.data$IS_RIGHT == T) %>%
      dplyr::select(-.data$IS_RIGHT)
  } else if (test == "skip") {
    result <-
      result %>%
      dplyr::filter(!is.na(.data$LBORRES) &
        is.na(.data$LBNRIND)) %>%
      dplyr::select(-.data$IS_RIGHT)
  } else if (test == "null") {
    result <-
      result %>%
      dplyr::filter(is.na(.data$LBORRES) &
        is.na(.data$LBNRIND)) %>%
      dplyr::select(-.data$IS_RIGHT)
  } else {
    stop("uknown parameter ", test)
  }

  result
}

#' Reshape the dataset to a long view
#'
#' @param dataset A data frame.
#' @param obj An object. Class lab.
#' @param row_file A data frame. A data frame with parameters.
#' @param part A character scalar. Prefixes or postfixes.
#'
#' @return A data frame. The part of the final result.
#'
to_long.lab <- function(obj, dataset, row_file, part) {
  # object's parameters
  id <- obj[["id"]]
  age <- obj[["age"]]
  sex <- obj[["sex"]]
  normal <- obj[["normal"]]
  abnormal <- obj[["abnormal"]]
  is_post <- obj[["is_post"]]
  obj_cl <- obj[["clsig"]]

  # laboratory's parameters
  lbtest <- row_file$LBTEST
  lbtestcd <- row_file$LBORRES
  lbnrind <- row_file$LBNRIND
  lbornrlo <- as.double(row_file$LBORNRLO)
  lbornrhi <- as.double(row_file$LBORNRHI)
  age_low <- as.double(row_file$AGELOW)
  age_high <- as.double(row_file$AGEHIGH)
  pattern_sex <- paste0("^", row_file$SEX, "$")

  # laboratory's parameter with prefix or postfix
  lborres <- ifelse(is_post, paste0(lbtestcd, part), paste0(part, lbtestcd))
  lbnrind <- ifelse(is_post, paste0(lbnrind, part), paste0(part, lbnrind))

  vars_rename <- c("LBORRES" = lborres, "LBNRIND" = lbnrind)

  # filter by age and sex
  by_age_sex <- dataset %>%
    dplyr::mutate(!!age := as.double(!!age)) %>%
    dplyr::filter(dplyr::between(!!age, age_low, age_high), grepl(pattern_sex, !!sex))

  # validate by reference values
  result <- by_age_sex %>%
    dplyr::mutate(LBTESTCD = lbtestcd, LBTEST = lbtest, VISIT = part, LBORNRLO = lbornrlo, LBORNRHI = lbornrhi) %>%
    dplyr::select(!!id, !!age, !!sex, .data$LBTEST, .data$LBTESTCD, .data$VISIT, .data$LBORNRLO, .data$LBORNRHI, !!lborres, !!lbnrind) %>%
    dplyr::mutate(RES_TYPE_NUM = to_dbl(.data[[lborres]])) %>%
    dplyr::mutate(IND_EXPECTED = ifelse(dplyr::between(.data$RES_TYPE_NUM, lbornrlo, lbornrhi), normal, abnormal)) %>%
    dplyr::mutate(IS_RIGHT = .data$IND_EXPECTED == .data[[lbnrind]]) %>%
    dplyr::rename(!!vars_rename)

  result
}
